### PR TITLE
 Add stub device plugin for e2e tests

### LINKED ISSUE #66952: stub device plugin based e2e test
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line:
>
> /kind bug
/kind feature

**What happened**:

**What you expected to happen**:
We need to have a cluster level stub device plugin based e2e test, which has been long pending task slip through several cycles. After we have that, we can configure it to run in different environments, like scale, soak, upgrade, and downgrade.

For simplicity, we can create a stub_device_plugin directory under test/images/ with a simple stub device plugin implementation in a .go file, and simple Makefile, Dockerfile, etc. The stub device plugin can start with defined behavior and just provide simple container runtime setting during Allocate call. For reference, pkg/kubelet/util/pluginwatcher/example_plugin.go is an example plugin we have used in pluginwatcher test. We can have a similar one for stub device plugin that implements the device plugin API.

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by vikaschoudhary16**
**What this PR does / why we need it**:
This PR adds:
- code for a sample device plugin
- Dockerfile and yaml files for the deployment of sample device plugin
- changes in the device plugin e2e-node test to use this device plugin

This same sample device plugin will be used in follow-up PRs to add e2e tests and run device plugin at cluster level.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Partially Fixes #66952 

**Special notes for your reviewer**:

**Release note**:

```release-note
None
```
/area hw-accelerators
/sig node 
/cc @jiayingz @yujuhong @dashpole @derekwaynecarr @vishh @RenaudWasTaken @ScorpioCPH @sjenning @FengyunPan2 @kubernetes/sig-node-pr-reviews

### CODE REVIEW

File: test/e2e/framework/device_plugin_util.go
Comment (dashpole) [inline test/e2e/framework/device_plugin_util.go]: Can we reference the yaml in `test/e2e/testing-manifests/sample-device-plugin.yaml` instead of this url?

File: test/e2e_node/device_plugin.go
Comment (dashpole) [inline test/e2e_node/device_plugin.go]: do we need to log all of the environment variables?

File: test/e2e_node/device_plugin.go
Comment (dashpole) [inline test/e2e_node/device_plugin.go]: should this be a constant?

File: test/images/sample-device-plugin/1
Comment (dashpole) [inline test/images/sample-device-plugin/1]: Do we need this file?  It looks like a copy of the sample device plugin below

**Review by dashpole — COMMENTED** [review]
looks great.  Just a couple small comments/questions

**Comment by dashpole** [comment]
I can help host this on k8s.gcr.io once we merge this.

File: test/e2e/framework/device_plugin_util.go
Comment (dashpole) [inline test/e2e/framework/device_plugin_util.go]: nit: leave a comment explaining where we are (pwd), and where the yaml is.

File: test/images/sample-device-plugin/1
Comment (dashpole) [reply test/images/sample-device-plugin/1]: bump

**Review by dashpole — COMMENTED** [review]
just one nit, otherwise lgtm

File: test/e2e/framework/device_plugin_util.go
Comment (vikaschoudhary16) [reply test/e2e/framework/device_plugin_util.go]: added comment for pwd. Regarding yaml, constant at L42 is showing path. Do we still need comment about yaml path?

Comment (dashpole) [reply test/e2e/framework/device_plugin_util.go]: this is good.  I just wanted to know what `/../../../../` is cd'ing out of.

File: test/e2e/testing-manifests/sample-device-plugin.yaml
Comment (dashpole) [inline test/e2e/testing-manifests/sample-device-plugin.yaml]: add a todo to move this to a k8s-owned repo

File: test/images/sample-device-plugin/sampledeviceplugin.go
Comment (jiayingz) [inline test/images/sample-device-plugin/sampledeviceplugin.go]: Not necessary in this PR, is it possible to skip Register gRPC call if kubelet supports DevicePluginProbe?

**Review by jiayingz — COMMENTED** [review]
/lgtm
Looking forward to the follow-up PR on the e2e test. Thanks a lot for the work!

File: test/images/sample-device-plugin/sampledeviceplugin.go
Comment (vikaschoudhary16) [reply test/images/sample-device-plugin/sampledeviceplugin.go]: Register() exits if DevicePluginProbe is supported. PTAL, https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/devicemanager/device_plugin_stub.go#L156

Comment (jiayingz) [reply test/images/sample-device-plugin/sampledeviceplugin.go]: Ah I forgot that part.

File: test/e2e/framework/device_plugin_util.go
Comment (RenaudWasTaken) [reply test/e2e/framework/device_plugin_util.go]: Nit: pwd -> cwd
Using the Gopath seems more reliable here (something like $GOPTAH/k8s.io/kubernetes)

File: test/e2e/testing-manifests/sample-device-plugin.yaml
Comment (RenaudWasTaken) [inline test/e2e/testing-manifests/sample-device-plugin.yaml]: Should we version the name (sample-beta-device-plugin)?

File: test/images/sample-device-plugin/sampledeviceplugin.go
Comment (RenaudWasTaken) [inline test/images/sample-device-plugin/sampledeviceplugin.go]: should we use glog?

File: test/images/sample-device-plugin/sampledeviceplugin.go
Comment (RenaudWasTaken) [inline test/images/sample-device-plugin/sampledeviceplugin.go]: Should there be some retry logic (retry 3 times at 5s interval) or is it unnecessary?

File: test/images/sample-device-plugin/sampledeviceplugin.go
Comment (RenaudWasTaken) [inline test/images/sample-device-plugin/sampledeviceplugin.go]: Use `select {}` it is less CPU intensive and is more idiomatic :)

File: test/images/sample-device-plugin/sampledeviceplugin.go
Comment (RenaudWasTaken) [inline test/images/sample-device-plugin/sampledeviceplugin.go]: error isn't handled here

**Review by RenaudWasTaken — COMMENTED** [review]
A few comments, mostly looks good.

File: test/e2e/framework/device_plugin_util.go
Comment (vikaschoudhary16) [reply test/e2e/framework/device_plugin_util.go]: looks correct to me :)

File: test/images/sample-device-plugin/sampledeviceplugin.go
Comment (vikaschoudhary16) [reply test/images/sample-device-plugin/sampledeviceplugin.go]: it would hardly run for nanosecs every 5 mins. That you want to save. cool :)

File: test/e2e/framework/device_plugin_util.go
Comment (RenaudWasTaken) [reply test/e2e/framework/device_plugin_util.go]: Sorry looks like my font changed the identation :(

File: test/images/sample-device-plugin/sampledeviceplugin.go
Comment (vikaschoudhary16) [reply test/images/sample-device-plugin/sampledeviceplugin.go]: @RenaudWasTaken I could not find a way where i could avoid `for` with `sleep`. If i use just `select{}`, it gives `fatal error: all goroutines are asleep - deadlock!`
```
package main

func main() {
    select {} // block forever
}
```

File: test/images/sample-device-plugin/sampledeviceplugin.go
Comment (vikaschoudhary16) [reply test/images/sample-device-plugin/sampledeviceplugin.go]: In the test we are verifying that kubelet is ready before starting plugin so seems kind of not adding much value.

File: test/images/sample-device-plugin/BASEIMAGE
Comment (mkumatag) [inline test/images/sample-device-plugin/BASEIMAGE]: lets add other platforms ref - https://github.com/kubernetes/kubernetes/blob/master/test/images/volume/nfs/BASEIMAGE

File: test/e2e/testing-manifests/sample-device-plugin.yaml
Comment (mkumatag) [inline test/e2e/testing-manifests/sample-device-plugin.yaml]: Lets build the image from `test/images/sample-device-plugins` and use that here!

/cc @ixdy

**Review by mkumatag — CHANGES_REQUESTED** [review]

File: test/e2e/testing-manifests/sample-device-plugin.yaml
Comment (vikaschoudhary16) [reply test/e2e/testing-manifests/sample-device-plugin.yaml]: @mkumatag build is created from `test/images/sample-device-plugins` only. As mentioned in comments, after this PR gets merged, image will be pushed to official registry and link would be updated here.

Comment (mkumatag) [reply test/e2e/testing-manifests/sample-device-plugin.yaml]: @vikaschoudhary16 usually, the image will be built and pushed first before any reference. We can even have a separate PR for just image and get it populated before the reference. @ixdy what do you think?

One more thing: please refer the multi-arch fat manifest instead of any `-arch` specific image.

Comment (vikaschoudhary16) [reply test/e2e/testing-manifests/sample-device-plugin.yaml]: @mkumatag 
>please refer the multi-arch fat manifest instead of any -arch specific image.

No idea what you mean by ^ . What is a "multi-arch fat manifest"? Can you share reference link please?

Comment (mkumatag) [reply test/e2e/testing-manifests/sample-device-plugin.yaml]: > Regarding image push, @dashpole suggested that we could push it post this PR merge.
> [#67928 (comment)](https://github.com/kubernetes/kubernetes/pull/67928#issuecomment-451611996)

Most of the folks first submit the code for the image, push images and then refer that here in the tests. That is the process been followed so far.

I'll leave the final decisions to @ixdy @luxas.

Comment (vikaschoudhary16) [reply test/e2e/testing-manifests/sample-device-plugin.yaml]: @mkumatag See [here](https://github.com/kubernetes/kubernetes/pull/67928/files#diff-92a015323f660c3ccafaf84bce3701a6R2). It is referring multiple platforms. If this is not what you meant, can you please share link to any such Dockerfile that is already being used in k8s?

File: test/images/sample-device-plugin/sampledeviceplugin.go
Comment (ixdy) [inline test/images/sample-device-plugin/sampledeviceplugin.go]: `example.com` might be better than `fake.com`?

File: test/images/sample-device-plugin/BASEIMAGE
Comment (ixdy) [inline test/images/sample-device-plugin/BASEIMAGE]: why centos?

File: test/e2e_node/device_plugin.go
Comment (ixdy) [inline test/e2e_node/device_plugin.go]: maybe also delete `stubAllocFunc` from this file?

File: test/images/sample-device-plugin/sampledeviceplugin.go
Comment (ixdy) [inline test/images/sample-device-plugin/sampledeviceplugin.go]: I know this is a copy-paste, but this docstring doesn't really describe what this is doing.

File: test/images/sample-device-plugin/sampledeviceplugin.go
Comment (ixdy) [inline test/images/sample-device-plugin/sampledeviceplugin.go]: shouldn't we abort/exit if there is an error rather than continuing silently?

Comment (ixdy) [reply test/images/sample-device-plugin/sampledeviceplugin.go]: also please use `klog` rather than `fmt.Println`.

File: test/images/sample-device-plugin/sampledeviceplugin.go
Comment (ixdy) [reply test/images/sample-device-plugin/sampledeviceplugin.go]: right, `select {}` will fail if there aren't any other running goroutines. I don't know what `dp1.Start()` and `dp1.Register()` do - do they start goroutines?

File: test/e2e/testing-manifests/sample-device-plugin.yaml
Comment (ixdy) [reply test/e2e/testing-manifests/sample-device-plugin.yaml]: @mkumatag is suggesting that this be `gcr.io/kubernetes-e2e-test-images/sample-device-plugin:1.0` with no -ARCH suffix.

I also agree that it might be worth separating the image and e2e test changes into two separate PRs.

Comment (vikaschoudhary16) [reply test/e2e/testing-manifests/sample-device-plugin.yaml]: @ixdy PTAL at seperate PR for image changes, https://github.com/kubernetes/kubernetes/pull/73511

File: test/e2e_node/device_plugin.go
Comment (vikaschoudhary16) [reply test/e2e_node/device_plugin.go]: done

File: test/images/sample-device-plugin/sampledeviceplugin.go
Comment (vikaschoudhary16) [reply test/images/sample-device-plugin/sampledeviceplugin.go]: `Start` starts device plugin grpc server in a goroutine. `Register` does not start a goroutine.

**Comment by ixdy** [comment]
now that #73511 is merged, you'll need to rebase this (and remove changes to `test/images`).

**Comment by dashpole** [comment]
I dont have permissions to push to `grc.io/kubernetes-e2e-test-images`.  I thought we were pushing this to `k8s.gcr.io`, which I could help with.  It looks like we will have to find someone else to build and push the image.

**Comment by ixdy** [comment]
@dashpole I've already pushed the image, but you should have access too. let's follow up on chat.

**Comment by mkumatag** [comment]
> @dashpole I've already pushed the image, but you should have access too. let's follow up on chat.

I see the images:

```shell
$ docker manifest inspect gcr.io/kubernetes-e2e-test-images/sample-device-plugin:1.0
{
   "schemaVersion": 2,
   "mediaType": "application/vnd.docker.distribution.manifest.list.v2+json",
   "manifests": [
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 739,
         "digest": "sha256:921ea299bd139de5ab26b647d5cb474aa8bb936e810af23f735ce88e166c6a98",
         "platform": {
            "architecture": "amd64",
            "os": "linux"
         }
      },
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 946,
         "digest": "sha256:7ca1d0eb721206994c6ac076e365cc7d9faa1403ce49f9b685b25c214e5cd03b",
         "platform": {
            "architecture": "arm",
            "os": "linux"
         }
      },
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 946,
         "digest": "sha256:558a62d42400f1a1bdefe88b8f20d5f9ed2c4ad1a0fd1816ed393424ba6534d0",
         "platform": {
            "architecture": "arm64",
            "os": "linux"
         }
      },
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 946,
         "digest": "sha256:8372974107d22ba1250ad3c2783beaea9b59fa83d622d6e1d6fe60e19a51bf5f",
         "platform": {
            "architecture": "ppc64le",
            "os": "linux"
         }
      },
      {
         "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
         "size": 946,
         "digest": "sha256:f887ee6216c044c9356a7a9e0fe444fc61b2321a3c9f82e4babd2a5cab162266",
         "platform": {
            "architecture": "s390x",
            "os": "linux"
         }
      }
   ]
}
```

**Comment by vikaschoudhary16** [comment]
thanks @mkumatag @ixdy @dashpole ! Updated image ref.

/hold cancel

**Comment by vikaschoudhary16** [comment]
ping @ixdy @mkumatag @dashpole

**Comment by dashpole** [comment]
This still /lgtm

File: test/e2e/framework/device_plugin_util.go
Comment (mkumatag) [inline test/e2e/framework/device_plugin_util.go]: This will create an issue if you want to try running the tests from e2e generated binary, you may need to include this yaml in the generated code and read from that.

**Review by mkumatag — CHANGES_REQUESTED** [review]

File: test/e2e/testing-manifests/sample-device-plugin.yaml
Comment (mkumatag) [inline test/e2e/testing-manifests/sample-device-plugin.yaml]: Why do you really have the same values from both request and limits? Do we really need them here for this testcase? If not please remove them, because I don't want to see these testcases failing on other platforms where memory allocation is little higher than x86

File: test/e2e/framework/device_plugin_util.go
Comment (oomichi) [inline test/e2e/framework/device_plugin_util.go]: nit: better to move "os" to line 21 for the same **standard** library import.

File: test/e2e/framework/device_plugin_util.go
Comment (oomichi) [inline test/e2e/framework/device_plugin_util.go]: Do SampleResourceName and SampleDevicePluginDSYAML need to be exposed?
They are just used in this module only, will we use them outside of this module with follow-up PRs?

File: test/e2e_node/device_plugin.go
Comment (oomichi) [inline test/e2e_node/device_plugin.go]: We can use `framework.ExpectNoError(err)` instead of `Expect(err).NotTo(BeNil())`

File: test/e2e/framework/device_plugin_util.go
Comment (oomichi) [inline test/e2e/framework/device_plugin_util.go]: nit: How about adding `sampleDevicePluginName` here like `sampleDevicePluginName = "sample-device-plugin"` and using it at line 58?

File: test/e2e/framework/device_plugin_util.go
Comment (oomichi) [inline test/e2e/framework/device_plugin_util.go]: How about `GetSampleDevicePluginPod()` ?

File: test/e2e_node/device_plugin.go
Comment (oomichi) [inline test/e2e_node/device_plugin.go]: line 64-73 can be replaced with `framework.WaitForNodeToBeReady()` call.

**Review by oomichi — COMMENTED** [review]
Thanks for updating, most part is good for me.
Just few comments

**Comment by oomichi** [comment]
Thanks for updating

/lgtm

**Comment by vikaschoudhary16** [comment]
@mkumatag @oomichi Thanks a lot for reviewing. Whom should i ping for approval? @ixdy ?

**Comment by vikaschoudhary16** [comment]
ping @mkumatag @oomichi @ixdy @dims 
Who else should i ping to get this approved?

**Review by timothysc — CHANGES_REQUESTED** [review]
Please move this to be a separate library under framework vs part of the core framework itself. 
It's only imported by a limited set of tests. 

Please see the existing patterns.

**Review by timothysc — APPROVED** [review]
There are minor nits. 
/approve 

^from an overall framework integration 

/hold 
^ I'll defer to @dashpole on final lgtm, but please address other comments.

File: test/e2e_node/device_plugin.go
Comment (vikaschoudhary16) [reply test/e2e_node/device_plugin.go]: addressed, changed to `ExpectError`

**Pull request merged**
